### DISCUSSION TITLE
Customizing Notification Features in Meshcentral

### DISCUSSION THREAD

**Discussion by amoljagdalepucsd** [Ylianst/MeshCentral#6005] [Q&A]
Greetings, community! In Meshcentral, we have a notify-user feature used to display notifications on remote devices, notifying users after a session starts. However, I aim to customize this feature by changing the default system icon to the Meshcentral logo or a custom icon, along with altering the window title (Windows Powershell to Meshcentral ) . After conducting some research, I've discovered that this customization requires modifications to the Meshagent source code.

Below is a snippet of code from the "toaster.JS" file, where notifications are formed on the remote system by executing a PowerShell script within a Node.js child process: we are loading system.windows.forms assembly .NET classes. I tried changing $objBalloon.Icon with different approaches and other scripts as well but did not work. 
Can someone please help me in this.

// Spawn a PowerShell process to set up the notification toast
retVal.child = require('child_process').execFile(process.env['windir'] + '\\System32\\WindowsPowerShell\\v1.0\\powershell.exe', ['powershell', '-noprofile', '-nologo', '-command', '-'], retVal.options);
retVal.child.descriptorMetadata = 'toaster';
retVal.child.toast = retVal;
retVal.child.stdout.stdin = retVal.child.stdin;
retVal.child.stderr.stdin = retVal.child.stdin;

// Handle toast dismissal event
retVal.child.stdout.on('data', function (c) { 
    if (c.toString().includes('<DISMISSED>')) { 
        this.stdin.write('exit\n'); 
    } 
});

// Set up PowerShell script for the notification
retVal.child.stderr.once('data', function (c) { 
    this.stdin.write('$objBalloon.dispose();exit\n'); 
});
**retVal.child.stdin.write('[System.Reflection.Assembly]::LoadWithPartialName("System.Windows.Forms")\r\n');
retVal.child.stdin.write('$objBalloon = New-Object System.Windows.Forms.NotifyIcon\r\n');
retVal.child.stdin.write('$objBalloon.Icon = [System.Drawing.SystemIcons]::Information\r\n');**
retVal.child.stdin.write('$objBalloon.Visible = $True\r\n');
retVal.child.stdin.write('Register-ObjectEvent -InputObject $objBalloon -EventName BalloonTipClosed -Action { $objBalloon.dispose();Write-Host "<`DISMISSED`>" }\r\n');
retVal.child.stdin.write('$objBalloon.ShowBalloonTip(10000, $env:_title, $env:_caption, 1)\r\n');

// Set a timeout to clean up after 10 seconds if the user doesn't interact with the toast
retVal.child.timeout = setTimeout(function (c) {
    c.timeout = null;
    c.stdin.write('$objBalloon.dispose();exit\n');
}, 10000, retVal.child);

// Handler for PowerShell process exit
retVal.child.on('exit', function () {
    if (this.timeout != null) { 
        clearTimeout(this.timeout); 
    }
    this.toast._res('DISMISSED');
});

return (retVal);

This is the screenshot : 
<img width="311" alt="notify" src="https://github.com/Ylianst/MeshCentral/assets/48832403/4f411c49-7dec-4352-a6a7-8ecca1f39fb7">

**Comment by si458** [comment]
you cant just change the JS files that are located inside the MeshAgent sourcecode.
you have to compile the exe with them, its a long process.

but you can PATCH the meshagent to see if things work before actually fully compiling the exe
https://ylianst.github.io/MeshCentral/meshcentral/debugging/#on-the-fly-patching-meshagent

**Reply by si458** [reply]
but also yes i agree, we should be able to pass an `SystemIcons` of some form into the `toaster` function to change the icon 👍

**Reply by amoljagdalepucsd** [reply]
Okay si458. I will go through the link and check if anything helps.
Also, we do not need to compile the .exe to see updated changes in modules (such as .js files) during development. Instead, we simply need to save the file in the same installed folder, for example, C:\Program Files\Meshagent. After making changes, restarting the Meshagent service from the task manager will reflect the updates. I can see the changes, but the only issue is that the icon and window title are not changing.

**Reply by si458** [reply]
@amoljagdalepucsd yes sorry ive just re-read the link and the info, you are correct.
the way i do it was slightly different, i would edit the code, then compress all modules into the libpolyfill, then compile and replace exe (the long way around haha)

**Reply by amoljagdalepucsd** [reply]
Okay sid458, I am trying it and will try to resolve it.
